src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,188 @@
 package core.basesyntax;
 
+import java.util.Collection;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        final Node<T> l = tail;
+        final Node<T> newNode = new Node<>(l, value, null);
+        tail = newNode;
+
+        if (l == null) {
+            head = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddingIndex(index);
+
+        if (index == size) {
+            add(value);
+        } else {
+            linkBefore(value, getNode(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    private void addAll(int index, Collection<? extends T> collection) {
+        checkAddingIndex(index);
+        T[] objects = (T[]) collection.toArray();
+
+        if (index == size) {
+            for (T elem : objects) {
+                add(elem);
+            }
+        } else {
+            for (T element : objects) {
+                linkBefore(element, getNode(index));
+                index++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> setNode = getNode(index);
+        final T oldValue = setNode.item;
+        setNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = head; current != null; current = current.next) {
+            if ((current.item != null && current.item.equals(object)) || current.item == object) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkBefore(T value, Node<T> current) {
+        if (current == null) {
+            throw new IndexOutOfBoundsException("Can't link element "
+                    + value + " before null Node " + current);
+        }
```

Comment:
```suggestion
```

Reply:
> Suggested change
>         if (current == null) {
>             throw new IndexOutOfBoundsException("Can't link element "
>                     + value + " before null Node " + current);
>         }

I really do not understand why we need to remove null-checking.